src/core/connection.c

```diff
@@ -5656,6 +5700,55 @@ QuicConnProcessUdpUnreachable(
     }
 }
 
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void
+QuicConnProcessRouteCompletion(
+    _In_ QUIC_CONNECTION* Connection,
+    _In_ const uint8_t* PhysicalAddress,
+    _In_ uint8_t PathId,
+    _In_ BOOLEAN Succeeded
+    )
+{
+    uint8_t PathIndex;
+    QUIC_PATH* Path = QuicConnGetPathByID(Connection, PathId, &PathIndex);
+    if (Path != NULL) {
+        if (Succeeded) {
+            CxPlatResolveRouteComplete(Connection, &Path->Route, PhysicalAddress, PathId);
+            QuicSendQueueFlush(&Connection->Send, REASON_ROUTE_COMPLETION);
+            QuicTraceLogConnInfo(
+                SuccessfulRouteResolution,
+                Connection,
+                "Processing successful route completion Path[%hhu]",
+                PathId);
+        } else {
+            //
+            // Kill the path that failed route resolution and make the next path active if possible.
+            //
+            if (Path->IsActive && Connection->PathsCount > 1) {
+                QuicTraceLogConnInfo(
+                    FailedRouteResolution,
+                    Connection,
+                    "Processing failed route completion Path[%hhu]",
+                    PathId);
+                QuicPathSetActive(Connection, &Connection->Paths[1]);
```

Comment:
`QuicPathSetActive` will change the order of the paths in the array. So it will invalidate `PathIndex` which you are using below when you call `QuicPathRemove`.

Reply:
It's exactly what the code expects. In that case, QuicPathRemove removes the path that just got swapped out.